test/integration/engine.go

```diff
@@ -0,0 +1,277 @@
+was not any synchronization. We use an exclusive flock(2) as a read lock, and
+obtain a shared lock as a read gate.
+
+Without this, tests would flake on a cold cache.
+*/
+
+func fetchArchive(t testing.TB) {
+	p, err := cachedDir()
+	if errors.Is(err, os.ErrNotExist) {
+		os.MkdirAll(p, 0755)
+	}
+	lf, err := os.Open(p)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer lf.Close()
+	defer syscall.Flock(int(lf.Fd()), syscall.LOCK_UN)
+	if err := syscall.Flock(int(lf.Fd()), syscall.LOCK_EX|syscall.LOCK_NB); err != nil {
+		// Failed to lock, wait for a shared lock, then return
+		t.Log("waiting for another process to fetch binaries")
+		if err := syscall.Flock(int(lf.Fd()), syscall.LOCK_SH); err != nil {
+			t.Fatal(err)
+		}
+		return
+	}
+
+	u := downloadURL()
+	t.Logf("fetching %q", u)
+	res, err := http.Get(u)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer res.Body.Close()
+	if res.StatusCode != http.StatusOK {
+		t.Fatalf("unexpected response: %v", res.Status)
+	}
+	t.Log("fetch OK")
+	tf, err := ioutil.TempFile(t.TempDir(), "embedded-postgres.")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tf.Close()
+	sz, err := io.Copy(tf, res.Body)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := tf.Seek(0, io.SeekStart); err != nil {
+		t.Fatal(err)
+	}
+	r, err := zip.NewReader(tf, sz)
+	if err != nil {
+		t.Fatal(err)
+	}
+	ex := false
+	for _, h := range r.File {
+		if !strings.HasSuffix(h.Name, ".txz") {
+			continue
+		}
+		t.Logf("extracting %q to %q", h.Name, p)
+		if err := os.MkdirAll(p, 0755); err != nil {
+			t.Error(err)
+		}
+		rd, err := h.Open()
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer rd.Close()
+		tf, err := xz.NewReader(rd)
+		if err != nil {
+			t.Fatal(err)
+		}
+		tr := tar.NewReader(tf)
+		var h *tar.Header
+		for h, err = tr.Next(); err == nil; h, err = tr.Next() {
+			outName := filepath.Join(p, h.Name)
+			switch h.Typeflag {
+			case tar.TypeDir:
+				if err := os.MkdirAll(outName, h.FileInfo().Mode()); err != nil {
+					t.Error(err)
+				}
+			case tar.TypeReg:
+				if err := os.MkdirAll(filepath.Dir(outName), 0755); err != nil {
+					t.Error(err)
+				}
+				f, err := os.Create(outName)
+				if err != nil {
+					t.Error(err)
+				}
+				// Defer error checking so we're unconditionally closing the
+				// file.
+				_, cerr := io.Copy(f, tr)
+				merr := f.Chmod(h.FileInfo().Mode())
+				if err := f.Close(); err != nil {
+					t.Error(err)
+				}
+				switch {
+				case cerr != nil:
+					t.Error(cerr)
+				case merr != nil:
+					t.Error(merr)
+				}
+			case tar.TypeSymlink:
+				if err := os.MkdirAll(filepath.Dir(outName), 0755); err != nil {
+					t.Error(err)
+				}
+				if err := os.Symlink(h.Linkname, outName); err != nil {
+					t.Error(err)
+				}
+			}
+			if t.Failed() {
+				t.FailNow()
+			}
+		}
+		if err != io.EOF {
+			t.Fatal(err)
+		}
+		ex = true
+		break
+	}
+	if !ex {
+		t.Fatal("didn't find txz")
+	}
+	t.Log("extraction OK")
+}
+
+func binUncached(t testing.TB) bool {
```

Comment:
Should we be returning `true` here if we get errors? We kind of have 3 (or more) states:

* Doesn't exist
* Exists and can get a lock (therefore fully available)
* Exists but error opening or getting the lock (I think flock will block until it gets the lock, but I guess the function can still return an error)

Reply:
I've got it logging the error and then failing the test. I'll add some explicit commentary.